File: tree_primer/tree_primer__binary_search_tree_exist.cs
// 実行時間 20ms
using static System.Console;
using static System.Linq.Enumerable;

internal class Program
{
    static void Main()
    {
        int[] nkr = ReadLine().Split().Select( int.Parse ).ToArray();
        var s = new System.Collections.Generic.SortedSet<int> { nkr[2] };

        foreach( int _ in Range( 0, nkr[0] - 1 ) ) {
            int[] a = ReadLine().Split().Select( int.Parse ).ToArray();

            s.Add( a[0] );
            s.Add( a[1] );
        }
        foreach( int _ in Range( 0, nkr[1] ) ) {
            WriteLine( s.Contains( int.Parse( ReadLine() ) ) ? "Yes" : "No" );
        }
        return;
    }
}

File: tree_primer/tree_primer__binary_search_tree_exist.py
# 実行時間 10ms
n, k, r = list(map(int, input().split()))
s = set()

s.add(r)
for _ in [0] * (n - 1):
    a, b = list(map(int, input().split()))

    s.add(a)
    s.add(b)
for _ in [0] * k:
    print("Yes" if int(input()) in s else "No")
